File: neetcode150/01_arrays_and_hashing/0049_group_anagrams.py
"""
=====================================================================
LeetCode 49 : Group Anagrams                                    (Medium)
https://leetcode.com/problems/group-anagrams/
Category   : Arrays & Hashing
---------------------------------------------------------------------
PROBLEM
    Given an array of strings, group the anagrams together.
---------------------------------------------------------------------
INTUITION
    Two words are anagrams iff their ORDER-INSENSITIVE fingerprint
    matches. Fingerprint options: (a) 26-letter count TUPLE, (b) the
    sorted word. Hash the fingerprint → anagrams share a bucket.
---------------------------------------------------------------------
APPROACH 1 — Count-tuple key (BEST for long words)
    Time  Complexity : O(n·k)   Space Complexity : O(n·k)
    (n words × k letters each)
    WHY tuple: Python tuples are IMMUTABLE and therefore HASHABLE —
    they can be dict keys. Lists are mutable → unhashable → rejected.
---------------------------------------------------------------------
APPROACH 2 — Sorted-string key
    Time  Complexity : O(n·k log k)   Space Complexity : O(n·k)
    Simpler to write; slower on long words (k log k per word).
=====================================================================
"""

from __future__ import annotations
import time
import tracemalloc
from collections import defaultdict
from typing import List


# =====================================================================
# APPROACH 1 : 26-count tuple fingerprint
# =====================================================================
class Solution_CountKey:
    """
    Purpose : Group all anagrams of the input words together.
    Inputs  : strs — list of words to group.
    Output  : list of groups; each group holds its anagrams.
    """

    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # fingerprint (tuple of 26 counts) → list of words in that bucket.
        # defaultdict(list) auto-creates the list on first key access.
        groups: dict[tuple, list] = defaultdict(list)

        # Fingerprint every word exactly once.
        for word in strs:
            # count[0..25] = occurrences of 'a'..'z' in this word.
            count = [0] * 26
            for ch in word:
                # ord(ch) - ord('a') maps 'a'→0 … 'z'→25.
                count[ord(ch) - ord('a')] += 1

            # list → tuple: tuples are hashable, lists are not.
            # (1,0,2,...) is the word's order-insensitive fingerprint.
            key = tuple(count)

            # All words sharing the fingerprint land in the same bucket.
            groups[key].append(word)

        # dict_values → list of lists, the required output shape.
        return list(groups.values())


# =====================================================================
# APPROACH 2 : sorted-word fingerprint
# =====================================================================
class Solution_SortedKey:
    """
    Purpose : Group all anagrams of the input words together.
    Inputs  : strs — list of words to group.
    Output  : list of groups; each group holds its anagrams.
    """

    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # sorted word → group (canonical form as dict key).
        groups: dict[str, list] = defaultdict(list)

        # Fingerprint every word once.
        for word in strs:
            # Sorting erases order: "eat","tea","ate" all become "aet".
            key = "".join(sorted(word))

            # Same sorted form ⟺ same letter multiset ⟺ same group.
            groups[key].append(word)

        # Flatten into the required output shape.
        return list(groups.values())


# =====================================================================
# BENCHMARK — time + peak memory for both approaches
# =====================================================================
if __name__ == "__main__":
    strs = ["eat", "tea", "tan", "ate", "nat", "bat",
            "listen", "silent", "enlist", "inlets"]

    tracemalloc.start()
    t0 = time.perf_counter()
    r1 = Solution_CountKey().groupAnagrams(strs)
    t1 = time.perf_counter()
    _, peak1 = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    tracemalloc.start()
    t2 = time.perf_counter()
    r2 = Solution_SortedKey().groupAnagrams(strs)
    t3 = time.perf_counter()
    _, peak2 = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    print(f"Approach 1 (count-tuple key) : {len(r1)} groups  time = {(t1 - t0) * 1e6:.2f} µs  peak-mem = {peak1} bytes")
    print(f"Approach 2 (sorted key)      : {len(r2)} groups  time = {(t3 - t2) * 1e6:.2f} µs  peak-mem = {peak2} bytes")
    print("PASS : both approaches agree." if len(r1) == len(r2) else "FAIL : approaches disagree.")

File: neetcode150/01_arrays_and_hashing/0049_group_anagrams.cpp
/*
 * =====================================================================
 * LeetCode 49 : Group Anagrams                                   (Medium)
 * https://leetcode.com/problems/group-anagrams/
 * Category   : Arrays & Hashing
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Given an array of strings, group the anagrams together. Any order
 *   of groups is accepted.
 * ---------------------------------------------------------------------
 * INTUITION
 *   Two words are anagrams iff their "fingerprint" is identical.
 *   A fingerprint must be ORDER-INSENSITIVE: either the sorted word
 *   ("eat"→"aet") or the 26-letter count tuple. Hash the fingerprint
 *   → all anagrams land in the same bucket. Classic hash-key trick.
 * ---------------------------------------------------------------------
 * APPROACH 1 — Count-tuple key (BEST for long words)
 *   Fingerprint = tuple of 26 letter counts (Python) / string of 26
 *   counts (C++). Computing it costs O(k) per word, k = word length.
 *   Time  Complexity : O(n·k)   — n words × k letters each.
 *   Space Complexity : O(n·k)   — the output groups + key storage.
 *   WHY tuple: Python tuples are HASHABLE (immutable) → valid dict
 *   keys; lists are not. C++ has no tuple hash → serialize counts
 *   into a fixed-length string "1,0,2,..." to get a comparable key.
 * ---------------------------------------------------------------------
 * APPROACH 2 — Sorted-string key
 *   Fingerprint = sorted(word). Sorting costs O(k log k) per word.
 *   Time  Complexity : O(n·k log k)   Space Complexity : O(n·k).
 *   Simpler to write; slower for long words. Both are accepted.
 * =====================================================================
 */

#include <vector>          // std::vector — the input and output containers
#include <string>          // std::string — words
#include <unordered_map>   // std::unordered_map — fingerprint → group
#include <algorithm>       // std::sort — approach 2 fingerprint
#include <chrono>          // timing for the benchmark
#include <iostream>        // std::cout — benchmark output
#include <sys/resource.h>  // getrusage — peak RAM measurement

using namespace std;

/* =====================================================================
 * APPROACH 1 : 26-count fingerprint key
 * ===================================================================== */
class Solution_CountKey {
public:
    /*
     * Purpose : Group all anagrams of the input words together.
     * Inputs  : strs — list of words to group.
     * Output  : vector of groups; each group is a list of anagrams.
     */
    vector<vector<string>> groupAnagrams(const vector<string>& strs) {

        // key (26 counts serialized) → list of words sharing that key.
        unordered_map<string, vector<string>> groups;

        // Fingerprint every word exactly once.
        for (const string& w : strs) {

            // count[0..25] = how many 'a'..'z' this word contains.
            int count[26] = {0};
            for (char c : w) {
                ++count[c - 'a'];   // 'a'→0 … 'z'→25, no hash function needed
            }

            // Serialize counts with commas so "1,0" ≠ "10" (unambiguous).
            string key;
            for (int i = 0; i < 26; ++i) {
                key += to_string(count[i]) + ",";
            }

            // All words with the same 26-counts land in the same bucket.
            groups[key].push_back(w);
        }

        // Flatten the hash table into the required "list of groups".
        vector<vector<string>> result;
        for (auto& kv : groups) {
            result.push_back(kv.second);
        }
        return result;
    }
};

/* =====================================================================
 * APPROACH 2 : sorted-word fingerprint key
 * ===================================================================== */
class Solution_SortedKey {
public:
    /*
     * Purpose : Group all anagrams of the input words together.
     * Inputs  : strs — list of words to group.
     * Output  : vector of groups; each group is a list of anagrams.
     */
    vector<vector<string>> groupAnagrams(const vector<string>& strs) {

        // sorted word → group (canonical form as the dict key).
        unordered_map<string, vector<string>> groups;

        // Fingerprint every word once.
        for (const string& w : strs) {

            // Sorting erases order: "eat","tea","ate" all become "aet".
            string key = w;
            sort(key.begin(), key.end());

            // Same sorted form ⟹ same letter multiset ⟹ same group.
            groups[key].push_back(w);
        }

        // Flatten groups into the required output shape.
        vector<vector<string>> result;
        for (auto& kv : groups) {
            result.push_back(kv.second);
        }
        return result;
    }
};

/* =====================================================================
 * BENCHMARK — time + peak RAM for both approaches
 * ===================================================================== */
int main() {
    // The official example plus one long-word stress case.
    vector<string> strs = {"eat", "tea", "tan", "ate", "nat", "bat",
                           "listen", "silent", "enlist", "inlets"};

    // --- Approach 1 ---
    struct rusage b1, a1;
    getrusage(RUSAGE_SELF, &b1);
    auto s1 = chrono::high_resolution_clock::now();
    auto r1 = Solution_CountKey().groupAnagrams(strs);
    auto e1 = chrono::high_resolution_clock::now();
    getrusage(RUSAGE_SELF, &a1);
    double us1 = chrono::duration_cast<chrono::nanoseconds>(e1 - s1).count() / 1000.0;
    long mem1 = (a1.ru_maxrss - b1.ru_maxrss) / 1024;

    // --- Approach 2 ---
    struct rusage b2, a2;
    getrusage(RUSAGE_SELF, &b2);
    auto s2 = chrono::high_resolution_clock::now();
    auto r2 = Solution_SortedKey().groupAnagrams(strs);
    auto e2 = chrono::high_resolution_clock::now();
    getrusage(RUSAGE_SELF, &a2);
    double us2 = chrono::duration_cast<chrono::nanoseconds>(e2 - s2).count() / 1000.0;
    long mem2 = (a2.ru_maxrss - b2.ru_maxrss) / 1024;

    cout << "Approach 1 (count-tuple key) : " << r1.size() << " groups"
         << "  time = " << us1 << " µs  peak-ram-delta = " << mem1 << " MB\n";
    cout << "Approach 2 (sorted key)      : " << r2.size() << " groups"
         << "  time = " << us2 << " µs  peak-ram-delta = " << mem2 << " MB\n";

    // Both must produce the same number of groups for this input.
    cout << ((r1.size() == r2.size()) ? "PASS : both approaches agree."
                                      : "FAIL : approaches disagree.")
         << "\n";
    return 0;
}